cylc/flow/parsec/validate.py

```diff
@@ -1136,23 +1137,25 @@ def _coerce_type(cls, value):
         return val
 
 
-# BACK COMPAT: BroadcastConfigValidator
-# The DB at 8.0.x stores Interval values as neither ISO8601 duration
-# string or DurationFloat. This has been fixed at 8.1.0, and
-# the following class acts as a bridge between fixed and broken.
-# url:
-#     https://github.com/cylc/cylc-flow/pull/5138
-# from:
-#    8.0.x
-# to:
-#    8.1.x
-# remove at:
-#    8.x
 class BroadcastConfigValidator(CylcConfigValidator):
     """Validate and Coerce DB loaded broadcast config to internal objects."""
     def __init__(self):
         CylcConfigValidator.__init__(self)
 
+    @classmethod
+    def coerce_str(cls, value, keys) -> str:
+        """Coerce value to a string.
+
+        Examples:
+            >>> BroadcastConfigValidator.coerce_str('abc#def', None)
+            'abc#def'
+        """
+        # Prevent ParsecValidator from assuming '#' means comments;
+        # '#' has valid uses in shell script such as parameter substitution
+        if isinstance(value, str) and '#' in value:
+            value = f'"{value}"'
```

Comment:
This doesn't work if the string is already quoted e.g:

```console
$ cylc broadcast one -n root -s 'script="sleep 30 # ja!"'
ERROR: [{'error': {'message': '(type=string) script = ""sleep 30 # ja!""', 'traceback': ['graphql.error.located_error.GraphQLLocatedError: (type=string) script = ""sleep 30 # ja!""\n']}}]
```

Maybe use `shlex.quote` which has logic to avoid surplus quoting.